File: hrcustomization_synergy/hrcustomization_synergy/report/employee_checkin_report/employee_checkin_report.py
import frappe
from frappe import _
from frappe.utils import getdate, formatdate


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    
    return columns, data


def get_columns():
    """Define the columns for the Employee Checkin Report"""
    columns = [
        {
            "label": _("Employee Code"),
            "fieldname": "employee",
            "fieldtype": "Link",
            "options": "Employee",
            "width": 120
        },
        {
            "label": _("Employee Name"),
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "width": 180
        },
        {
            "label": _("Date"),
            "fieldname": "attendance_date",
            "fieldtype": "Date",
            "width": 100
        },
        {
            "label": _("Time In"),
            "fieldname": "in_time",
            "fieldtype": "Time",
            "width": 100
        },
        {
            "label": _("Time Out"),
            "fieldname": "out_time",
            "fieldtype": "Time",
            "width": 100
        },
        {
            "label": _("Shift"),
            "fieldname": "shift",
            "fieldtype": "Link",
            "options": "Shift Type",
            "width": 120
        },
        {
            "label": _("Status"),
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("Working Hours"),
            "fieldname": "working_hours",
            "fieldtype": "Float",
            "width": 100,
            "precision": 2
        },
        {
            "label": _("Expected Working Hours"),
            "fieldname": "expected_working_hours",
            "fieldtype": "Float",
            "width": 140,
            "precision": 2
        },
    ]
    
    return columns


def get_data(filters):
    """Fetch attendance data with employee checkin details"""
    from frappe.utils import getdate, time_diff_in_hours
    
    # Get attendance records
    conditions = get_conditions(filters)
    
    attendance_data = frappe.db.sql(f"""
        SELECT 
            a.employee,
            e.employee_name,
            a.attendance_date,
            a.shift,
            a.status,
            a.working_hours
        FROM `tabAttendance` a
        LEFT JOIN `tabEmployee` e ON a.employee = e.name
        WHERE a.docstatus = 1 {conditions}
        ORDER BY a.attendance_date DESC, a.employee
    """, as_dict=True)
    
    total_working_hours = 0
    total_expected_hours = 0
    
    # For each attendance record, get the first and last checkin times
    for record in attendance_data:
        # Get all checkins for this employee on this date
        checkins = frappe.db.sql("""
            SELECT 
                time
            FROM `tabEmployee Checkin`
            WHERE employee = %(employee)s
                AND DATE(time) = %(date)s
            ORDER BY time
        """, {
            'employee': record['employee'],
            'date': record['attendance_date']
        }, as_dict=True)
        
        if checkins:
            # First checkin is time in
            record['in_time'] = checkins[0]['time'].time() if checkins[0]['time'] else None
            # Last checkin is time out (if more than one checkin)
            record['out_time'] = checkins[-1]['time'].time() if len(checkins) > 1 and checkins[-1]['time'] else None
        else:
            record['in_time'] = None
            record['out_time'] = None
        
        # Expected working hours: 8 hours per day, 0 on Holiday
        expected_hours = 0 if record.get("status") == "Holiday" else 8
        record['expected_working_hours'] = expected_hours
        
        total_working_hours += record.get('working_hours') or 0
        total_expected_hours += expected_hours
    
    # Append total row at the end
    if attendance_data:
        attendance_data.append({
            "employee": "",
            "employee_name": _("Total"),
            "attendance_date": None,
            "in_time": None,
            "out_time": None,
            "shift": None,
            "status": "",
            "working_hours": total_working_hours,
            "expected_working_hours": total_expected_hours
        })
    
    return attendance_data


def get_conditions(filters):
    """Build SQL conditions based on filters"""
    conditions = ""
    
    if filters.get("employee"):
        conditions += f" AND a.employee = '{filters.get('employee')}'"
    
    if filters.get("from_date"):
        conditions += f" AND a.attendance_date >= '{filters.get('from_date')}'"
    
    if filters.get("to_date"):
        conditions += f" AND a.attendance_date <= '{filters.get('to_date')}'"
    
    if filters.get("department"):
        conditions += f" AND e.department = '{filters.get('department')}'"
    
    
    if filters.get("shift"):
        conditions += f" AND a.shift = '{filters.get('shift')}'"
    
    if filters.get("status"):
        conditions += f" AND a.status = '{filters.get('status')}'"
    
    return conditions

File: hrcustomization_synergy/hrcustomization_synergy/report/employee_checkin_report/employee_checkin_report.js
// Copyright (c) 2026, NexTash and contributors
// For license information, please see license.txt

frappe.query_reports["Employee Checkin Report"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "employee",
            "label": __("Employee"),
            "fieldtype": "Link",
            "options": "Employee"
        },
        {
            "fieldname": "department",
            "label": __("Department"),
            "fieldtype": "Link",
            "options": "Department"
        },
        {
            "fieldname": "project",
            "label": __("Project"),
            "fieldtype": "Link",
            "options": "Project"
        },
        {
            "fieldname": "shift",
            "label": __("Shift"),
            "fieldtype": "Link",
            "options": "Shift Type"
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": "\nPresent\nAbsent\nHalf Day\nWork From Home\nOn Leave"
        }
    ],
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        if (column.fieldname == "status" && data) {
            if (data.status === "Present") {
                value = "<span style='color: green'>" + value + "</span>";
            } else if (data.status === "Absent") {
                value = "<span style='color: red'>" + value + "</span>";
            } else if (data.status === "Half Day") {
                value = "<span style='color: orange'>" + value + "</span>";
            } else if (data.status === "Work From Home") {
                value = "<span style='color: blue'>" + value + "</span>";
            } else if (data.status === "On Leave") {
                value = "<span style='color: purple'>" + value + "</span>";
            }
        }
        
        if (column.fieldname == "coordinates" && data && data.coordinates) {
            // Make coordinates clickable to open in maps
            const coords = data.coordinates.split(', ');
            if (coords.length === 2) {
                const lat = coords[0].trim();
                const lng = coords[1].trim();
                const mapUrl = `https://www.google.com/maps?q=${lat},${lng}`;
                value = `<a href="${mapUrl}" target="_blank" title="View on Map">${data.coordinates}</a>`;
            }
        }
        
        return value;
    },
    
    onload: function(report) {
        // Add export functionality
        report.page.add_menu_item(__("Export to Excel"), function() {
            frappe.utils.csvify(report.data, report.columns, __("Employee Checkin Report"));
        });
        
        // Add summary statistics
        report.page.add_action_item(__("Show Summary"), function() {
            show_attendance_summary(report.data);
        });
    }
};

function show_attendance_summary(data) {
    if (!data || data.length === 0) {
        frappe.msgprint(__("No data to summarize"));
        return;
    }
    
    const summary = {
        total_records: data.length,
        present: 0,
        absent: 0,
        half_day: 0,
        work_from_home: 0,
        on_leave: 0,
        total_working_hours: 0,
        with_coordinates: 0
    };
    
    data.forEach(function(row) {
        if (row.status === "Present") summary.present++;
        else if (row.status === "Absent") summary.absent++;
        else if (row.status === "Half Day") summary.half_day++;
        else if (row.status === "Work From Home") summary.work_from_home++;
        else if (row.status === "On Leave") summary.on_leave++;
        
        if (row.working_hours) {
            summary.total_working_hours += row.working_hours;
        }
        
        if (row.coordinates && row.coordinates.trim()) {
            summary.with_coordinates++;
        }
    });
    
    const summary_html = `
        <div class="row">
            <div class="col-md-6">
                <h5>Attendance Summary</h5>
                <table class="table table-bordered table-condensed">
                    <tr><td>Total Records</td><td>${summary.total_records}</td></tr>
                    <tr><td style="color: green">Present</td><td>${summary.present}</td></tr>
                    <tr><td style="color: red">Absent</td><td>${summary.absent}</td></tr>
                    <tr><td style="color: orange">Half Day</td><td>${summary.half_day}</td></tr>
                    <tr><td style="color: blue">Work From Home</td><td>${summary.work_from_home}</td></tr>
                    <tr><td style="color: purple">On Leave</td><td>${summary.on_leave}</td></tr>
                </table>
            </div>
            <div class="col-md-6">
                <h5>Other Statistics</h5>
                <table class="table table-bordered table-condensed">
                    <tr><td>Total Working Hours</td><td>${summary.total_working_hours.toFixed(2)}</td></tr>
                    <tr><td>Avg Hours/Day</td><td>${(summary.total_working_hours / Math.max(summary.present + summary.half_day, 1)).toFixed(2)}</td></tr>
                    <tr><td>Records with GPS</td><td>${summary.with_coordinates}</td></tr>
                    <tr><td>GPS Coverage</td><td>${((summary.with_coordinates / summary.total_records) * 100).toFixed(1)}%</td></tr>
                </table>
            </div>
        </div>
    `;
    
    frappe.msgprint({
        title: __("Attendance Summary"),
        message: summary_html,
        wide: true
    });
}